kernel/src/metaptr.rs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the MetaPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata.
+/// This should be used any time the kernel wishes to grant authority to the user, or any time
+/// the user should be required to prove validity of a pointer.
+#[derive(Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
+pub struct MetaPtr {
+    ptr: usize,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum MetaPermissions {
+    Any,
+    Read,
+    Write,
+    ReadWrite,
+    Execute,
+}
+
+impl From<MetaPtr> for usize {
```

Comment:
I think this is OK because a `MetaPtr` is always convertible into a raw pointer by just losing any meta information. However, might be better to implement this for `<T> *const T` (which can always be cast to a `usize` anyway)?

Reply:
I can add that too (it might even exist as a method?), but this one gets used a lot in syscall / return argument handling.

Reply:
I think its happening a lot because values that are usize, or even eventually u32, are being passed in registers that can hold something as wide as `MetaPtr` (the widest type, or so I have assumed).

@bradjc has suggested that maybe those should be a separate `RegisterData` type, something large enough to hold `u32`/`isize`/`*const()`/`MetaPtr`.

I suppose, fairly, one could imagine a platform with pointers that are smaller than u32, in which case really RegisterData would wrap around the largest type.

Reply:
Documenting that `CapabilityPtr` is a pointer then using it to store other things is very confusing, and I think having a separate `RegisterData` [1] type would help that confusion. `CapabilityPtr` is already a confusing enough type to understand, especially for contributors who are unfamiliar with CHERI.

The fact that some CHERI systems use different registers for integers and capabilities however, does throw a wrench into the works. However, there are two obvious ways that come to mind for how we might choose to handle that at the ABI level:

1. We can pass integers in the integer registers and capabilities in the capability registers. In this case, the kernel knows which are which so we can can use `usize` for the integer registers and `CapabilityPtr` for the capability registers [2].
1. We can pass everything in the capability registers and ignore the integer registers. This is just treating the system as a CHERI system that uses the same registers for everything, and we'd just use `RegisterData` for all the register values.

In both cases, it makes sense to only use `CapabilityPtr` for pointer types, so I don't think it's really a blocker for having both `CapabilityPtr` and `RegisterData`. Option 1 would be somewhat awkward, though, because we would only use `RegisterData` for the capability registers.

[1] libtock-rs just calls this `Register`
[2] This ignores the possibility of non-pointer capabilities, which is another complication

Reply:
I am not convinced `RegisterData` is really a single type. Machines have different width registers. To say that `RegisterData` should refer to just one of them is odd to me. If we did introduce a `RegisterData` here, I would expect it to wrap `CapabilityPtr` (with an appropriate comment saying the type is being used because it is expected to be widest). It would certainly throw a spanner if ever a different ABI were introduced that actually used different width registers, to have a `RegisterData` and then `SomeOtherRegisterData`. But I am happy to add `RegisterData` if that is the consensus and makes the syscall layer easier to understand.

1. is/was the ABI for those split file machines. In a similar fashion to how you would select float/integer registers on platforms that split them.